src/main/java/io/swagger/codegen/languages/kotlin/AbstractKotlinCodegen.java

```diff
@@ -0,0 +1,573 @@
+        }
+        return super.toInstantiationType(p);
+    }
+
+    /**
+     * Return the fully-qualified "Model" name for import
+     *
+     * @param name the name of the "Model"
+     * @return the fully-qualified "Model" name for import
+     */
+    @Override
+    public String toModelImport(String name) {
+        // toModelImport is called while processing operations, but DefaultCodegen doesn't
+        // define imports correctly with fully qualified primitives and models as defined in this generator.
+        if (needToImport(name)) {
+            return super.toModelImport(name);
+        }
+
+        return name;
+    }
+
+    /**
+     * Output the proper model name (capitalized).
+     * In case the name belongs to the TypeSystem it won't be renamed.
+     *
+     * @param name the name of the model
+     * @return capitalized model name
+     */
+    @Override
+    public String toModelName(final String name) {
+        // Allow for explicitly configured kotlin.* and java.* types
+        if (name.startsWith("kotlin.") || name.startsWith("java.")) {
+            return name;
+        }
+
+        // If importMapping contains name, assume this is a legitimate model name.
+        if (importMapping.containsKey(name)) {
+            return importMapping.get(name);
+        }
+
+        String modifiedName = name.replaceAll("\\.", "");
+        modifiedName = sanitizeKotlinSpecificNames(modifiedName);
+
+        if (reservedWords.contains(modifiedName)) {
+            modifiedName = escapeReservedWord(modifiedName);
+        }
+
+        return titleCase(modifiedName);
+    }
+
+    @Override
+    public String toVarName(String name) {
+        return super.toVarName(sanitizeKotlinSpecificNames(name));        
+    }
+    
+    /**
+     * Provides a strongly typed declaration for simple arrays of some type and arrays of arrays of some type.
+     *
+     * @param arr
+     * @return
+     */
+    private String getArrayTypeDeclaration(ArraySchema arraySchema) {
+        // TODO: collection type here should be fully qualified namespace to avoid model conflicts
+        // This supports arrays of arrays.
+        String arrayType = typeMapping.get("array");
+        StringBuilder instantiationType = new StringBuilder(arrayType);
+        Schema items = arraySchema.getItems();
+        String nestedType = getTypeDeclaration(items);
+        // TODO: We may want to differentiate here between generics and primitive arrays.
+        instantiationType.append("<").append(nestedType).append(">");
+        return instantiationType.toString();
+    }    
+
+    /**
+     * Sanitize against Kotlin specific naming conventions, which may differ from those required by {@link DefaultCodegen#sanitizeName}.
+     *
+     * @param name string to be sanitize
+     * @return sanitized string
+     */
+    private String sanitizeKotlinSpecificNames(final String name) {
+        String word = removeNonNameElementToCamelCase(name);
```

Comment:
This might cause problems. Whatever `removeNonNameElementToCamelCase` does, if it changes `name` at all, it may make enums stop working.

For instance, if an enum is `last-call`, this method previously would have converted this to `last_call`. If `removeNonNameElementToCamelCase` changes `name` to `lastCall` or `lastcall`, that expectation won't be met. I couldn't find the implementation of `removeNonNameElementToCamelCase`, so I can't say for certain.